staging/src/k8s.io/component-base/metrics/prometheus/controller/metrics.go

```diff
@@ -0,0 +1,151 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"sync"
+	"time"
+
+    "k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+	"k8s.io/klog/v2"
+)
+
+type sinceFunc func(time.Time) time.Duration
+
+const controllerNamespace = "controller"
+
+var (
+	ReconciliationDurations = metrics.NewHistogramVec(
+		&metrics.HistogramOpts{
+			Namespace:      controllerNamespace,
+			Name:           "reconcile_time_seconds",
+			Help:           "Duration of a sync of a controller's resource",
+			StabilityLevel: metrics.ALPHA,
+			Buckets:        metrics.ExponentialBuckets(0.001, 2, 15),
+		},
+		[]string{"controller", "status"},
+	)
+
+	since sinceFunc = time.Since
+)
+
+var registerMetrics sync.Once
+
+// RegisterMetrics registers the controller metrics
+func RegisterMetrics() {
+	registerMetrics.Do(func() {
+		legacyregistry.MustRegister(ReconciliationDurations)
+	})
+}
+
+// SyncAndRecord wraps syncFn with a function, recording its result and duration as a metric.
+func SyncAndRecord(
+	controllerName string, syncFn func(context.Context, string) error) func(context.Context, string) error {
+	return func(ctx context.Context, key string) error {
+		return RunSyncAndRecord(ctx, controllerName, key, syncFn)
+	}
+}
+
+// RunSyncAndRecord runs syncFn with the given context and key, recording the reconcile time and results
+func RunSyncAndRecord(
+	ctx context.Context, controllerName string, key string, syncFn func(context.Context, string) error) error {
+	startTime := time.Now()
+	err := syncFn(ctx, key)
+	recordSyncResult(controllerName, startTime, err)
+	return err
+}
+
+// RunSyncAndRecordNoCtx runs syncFn with key, recording the reconcile time and results
+func RunSyncAndRecordNoCtx(
+	controllerName string, key string, syncFn func(string) error) error {
+	startTime := time.Now()
+	err := syncFn(key)
+	recordSyncResult(controllerName, startTime, err)
+	return err
+}
+
+// SyncAndRecordWithBool wraps syncFn with a function, recording its result and duration as a metric.
+func SyncAndRecordWithBool(
+	controllerName string, syncFn func(context.Context, string) (bool, error)) func(context.Context, string) (bool, error) {
+	return func(ctx context.Context, key string) (bool, error) {
+		return RunSyncAndRecordWithBool(ctx, controllerName, key, syncFn)
+	}
+}
+
+// SyncAndRecordWithBoolNoCtx wraps syncFn with a function, recording its result and duration as a metric.
+func SyncAndRecordWithBoolNoCtx(
```

Comment:
I would prefer `HasSyncedAndRecorded` and `HasSyncedAndRecordedWithCtx` rather than a `NoCtx` variant.